src/main/java/net/minecraftforge/client/model/renderable/SimpleRenderable.java

```diff
@@ -0,0 +1,183 @@
+
+    private SimpleRenderable()
+    {
+    }
+
+    public static Builder builder()
+    {
+        return new Builder();
+    }
+
+    @Override
+    public void render(PoseStack poseStack, MultiBufferSource bufferSource, Function<ResourceLocation, RenderType> renderTypeFunction, int lightmapCoord, int overlayCoord, float partialTicks, MultipartTransforms renderValues)
+    {
+        for(var part : parts)
+        {
+            part.render(poseStack, bufferSource, renderTypeFunction, lightmapCoord, overlayCoord, renderValues);
+        }
+    }
+
+    private static class Part
+    {
+        private final String name;
+        private final List<Part> parts = new ArrayList<>();
+        private final List<Mesh> meshes = new ArrayList<>();
+
+        public Part(String name)
+        {
+            this.name = name;
+        }
+
+        public void render(PoseStack poseStack, MultiBufferSource bufferSource, Function<ResourceLocation, RenderType> renderTypeFunction, int lightmapCoord, int overlayCoord, MultipartTransforms renderValues)
+        {
+            Matrix4f matrix = renderValues.getPartValues(name);
+            if (matrix != null)
+            {
+                poseStack.pushPose();
+                poseStack.mulPoseMatrix(matrix);
+            }
+
+            for(var part : parts)
+            {
+                part.render(poseStack, bufferSource, renderTypeFunction, lightmapCoord, overlayCoord, renderValues);
+            }
+
+            for(var mesh : meshes)
+            {
+                mesh.render(poseStack, bufferSource, renderTypeFunction, lightmapCoord, overlayCoord, renderValues);
+            }
+
+            if (matrix != null)
+            {
+                poseStack.popPose();
+            }
+        }
+    }
+
+    private static class Mesh
+    {
+        private final ResourceLocation texture;
+        private final List<BakedQuad> quads = new ArrayList<>();
+
+        public Mesh(ResourceLocation texture)
+        {
+            this.texture = texture;
+        }
+
+        public void render(PoseStack poseStack, MultiBufferSource bufferSource, Function<ResourceLocation, RenderType> renderTypeFunction, int lightmapCoord, int overlayCoord, MultipartTransforms renderValues)
+        {
+            var consumer = bufferSource.getBuffer(renderTypeFunction.apply(texture));
+            for(var quad : quads)
+            {
+                consumer.putBulkData(poseStack.last(), quad, 1, 1, 1, 1, lightmapCoord, overlayCoord, true);
+            }
+        }
+    }
+
+    public static class Builder
+    {
+        private final SimpleRenderable renderable = new SimpleRenderable();
+        private final Stack<Part> parts = new Stack<>();
+
+        @Nullable
+        private Part currentPart;
+
+        private List<Part> parentList;
+
+        public Builder()
+        {
+            parentList = renderable.parts;
+        }
+
+        public Builder pushPart(String partName)
```

Comment:
This concept feels iffy to me. Always has.

I generally prefer it if the part builder is either his own instance and not the same instance that is prefixing or even an instance that support auto closing so that a try with resources block could be used to wrap the part construction and handle the build code automatically

Reply:
I made it that way to avoid allocating extra objects, but I could switch to individual builders. I'll play around with it.

Reply:
Personally allocating the additional objects should not matter here.
The renderer should cache the renderable anyway, right?